test/test_charm.py

```diff
@@ -12,188 +12,155 @@
-                self.seen: typing.List[str] = []
-                for rel in ('req1', 'req-2', 'pro1', 'pro-2', 'peer1', 'peer-2'):
-                    # Hook up relation events to generic handler.
-                    self.framework.observe(self.on[rel].relation_joined, self.on_any_relation)
-                    self.framework.observe(self.on[rel].relation_changed, self.on_any_relation)
-                    self.framework.observe(self.on[rel].relation_departed, self.on_any_relation)
-                    self.framework.observe(self.on[rel].relation_broken, self.on_any_relation)
-
-            def on_any_relation(self, event: ops.RelationEvent):
-                assert event.relation.name == 'req1'
-                assert event.relation.app is not None
-                assert event.relation.app.name == 'remote'
-                self.seen.append(type(event).__name__)
+    framework = create_framework(request)
 
-        # language=YAML
-        self.meta = ops.CharmMeta.from_yaml(metadata='''
+    events: typing.List[str] = list(MyCharm.on.events())  # type: ignore
+    assert 'install' in events
+    assert 'custom' in events
+
+    charm = MyCharm(framework)
+    charm.on.start.emit()
+
+    assert charm.started
+
+    with pytest.raises(TypeError):
+        framework.observe(charm.on.start, charm)  # type: ignore
+
+
+def test_observe_decorated_method(request: pytest.FixtureRequest):
+    # we test that charm methods decorated with @functools.wraps(wrapper)
+    # can be observed by Framework. Simpler decorators won't work because
+    # Framework searches for __self__ and other method things; functools.wraps
+    # is more careful and it still works, this test is here to ensure that
+    # it keeps working in future releases, as this is presently the only
+    # way we know of to cleanly decorate charm event observers.
+    events: typing.List[ops.EventBase] = []
+
+    def dec(fn: typing.Any) -> typing.Callable[..., None]:
+        # simple decorator that appends to the nonlocal
+        # `events` list all events it receives
+        @functools.wraps(fn)
+        def wrapper(charm: 'MyCharm', evt: ops.EventBase):
+            events.append(evt)
+            fn(charm, evt)
+        return wrapper
+
+    class MyCharm(ops.CharmBase):
+        def __init__(self, *args: typing.Any):
+            super().__init__(*args)
+            framework.observe(self.on.start, self._on_start)
+            self.seen = None
+
+        @dec
+        def _on_start(self, event: ops.EventBase):
+            self.seen = event
+
+    framework = create_framework(request)
+    charm = MyCharm(framework)
+    charm.on.start.emit()
+    # check that the event has been seen by the decorator
+    assert len(events) == 1
+    # check that the event has been seen by the observer
+    assert isinstance(charm.seen, ops.StartEvent)
+
+
+def test_observer_not_referenced_warning(
+        caplog: pytest.LogCaptureFixture,
+        request: pytest.FixtureRequest):
```

Comment:
Nit: why don't we standardise on the `request` parameter always coming first, for a bit of consistency? It seems like it should go first because the "test context" (request) is kind of primary.